test/EndToEndTests/Common/Microsoft.OData.Client.E2E.TestCommon/Common/CustomHttpClientRequestMessage.cs

```diff
@@ -0,0 +1,127 @@
+﻿//-----------------------------------------------------------------------------
+// <copyright file="CustomHttpClientRequestMessage.cs" company=".NET Foundation">
+//      Copyright (c) .NET Foundation and Contributors. All rights reserved.
+//      See License.txt in the project root for license information.
+// </copyright>
+//------------------------------------------------------------------------------
+
+namespace Microsoft.OData.Client.E2E.TestCommon.Common
+{
+    /// <summary>
+    /// An implementation of IODataRequestMessageAsync that uses an HttpRequestMessage under the covers.
+    /// In ODataLibrary, a message is an abstraction which consists of stream and header interfaces that hides the details of stream-reading/writing.
+    /// </summary>
+    public class CustomHttpClientRequestMessage : IODataRequestMessageAsync, IServiceCollectionProvider
+    {
+        private readonly HttpRequestMessage _request;
+        private readonly HttpClient _httpClient;
+        private HttpContent _content;
+        private MemoryStream _memoryStream;
+
+        public CustomHttpClientRequestMessage(Uri uri, HttpClient httpClient)
+        {
+            _httpClient = httpClient;
+            _memoryStream = new MemoryStream();
+
+            _request = new HttpRequestMessage
+            {
+                RequestUri = uri
+            };
+        }
+
+        public string? GetHeader(string headerName)
+        {

```

Comment:
IMO, the implementation adopted in `HttpClientResponseMessage` class is clearer hence easier to follow/understand - https://github.com/OData/odata.net/pull/3031/files#diff-2702b14dda9b37a0d556e52920c0d2e93821e78e5407304cd2b38f10932a3d9fR23
I'd advise that you implement this method the same way in both classes